File: scala/secret-handshake/src/main/scala/SecretHandshake.scala
object SecretHandshake {
  private object Flags {
    val Wink         : Int = 1 << 0
    val DoubleBlink  : Int = 1 << 1
    val CloseYourEyes: Int = 1 << 2
    val Jump         : Int = 1 << 3
    val Reverse      : Int = 1 << 4
  }

  private val events = Seq(
    Flags.Wink -> "wink",
    Flags.DoubleBlink -> "double blink",
    Flags.CloseYourEyes -> "close your eyes",
    Flags.Jump -> "jump"
  )

  def commands(mask: Int): List[String] = {
    val handshake = events
      .filter { case (flag, _) => isSatisfied(flag, mask) }
      .map { case (_, event) => event }
      .toList

    if (isSatisfied(Flags.Reverse, mask)) handshake.reverse else handshake
  }

  private def isSatisfied(flag: Int, mask: Int): Boolean = (mask & flag) > 0
}

File: fsharp/secret-handshake/SecretHandshake.fs
module SecretHandshake

type Command =
    | Wink = 1
    | DoubleBlink = 2
    | CloseYourEyes = 4
    | Jump = 8
    | Reverse = 16

let (|HasFlag|_|) (flag: Command) n =
    if n &&& (int flag) > 0 then
        Some HasFlag
    else
        None

let commands number =
    let isActive command =
        match number with
        | HasFlag command -> true
        | _ -> false

    let activeCommands =
        [ Command.Wink, "wink"
          Command.DoubleBlink, "double blink"
          Command.CloseYourEyes, "close your eyes"
          Command.Jump, "jump" ]
        |> List.filter (fst >> isActive)
        |> List.map snd

    match number with
    | HasFlag Command.Reverse -> List.rev activeCommands
    | _ -> activeCommands
